File: samples/CslaLight/cs/Rolodex/Application/Helpers/RankConverter.cs
using System;
using System.Net;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Documents;
using System.Windows.Ink;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Media.Animation;
using System.Windows.Shapes;
using System.Windows.Data;
using System.Globalization;
using Rolodex.Business.BusinessClasses;

namespace Rolodex
{
  public class RankConverter : IValueConverter
  {
    Ranks rankList;
    public RankConverter()
    {
      Ranks.GetRanks((o, e) =>
        {
          rankList = e.Object;
          OnGotData();
        });
    }

    public event EventHandler GotData;

    protected void OnGotData()
    {
      if (GotData != null)
        GotData.Invoke(this, EventArgs.Empty);
    }

    public object Convert(object value, Type targetType, object parameter, System.Globalization.CultureInfo culture)
    {
      if (value != null && value is int && rankList != null)
      {
        return rankList.GetRankName((int)value);
      }
      else
        return string.Empty;;
    }

    public object ConvertBack(object value, Type targetType, object parameter, System.Globalization.CultureInfo culture)
    {
      if (value != null && value is string && rankList != null)
      {
        var returnValue = rankList.GetItemByValue((string)value);
        if (returnValue != null)
          return returnValue.Key;
        else
          return 0;
      }
      else
        return 0;
    }
  }
}

File: Samples/CslaLight/vb/Rolodex/Application/Helpers/RankConverter.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Net
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Documents
Imports System.Windows.Ink
Imports System.Windows.Input
Imports System.Windows.Media
Imports System.Windows.Media.Animation
Imports System.Windows.Shapes
Imports System.Windows.Data
Imports System.Globalization
Imports Rolodex.Business.BusinessClasses

Namespace Rolodex
  Public Class RankConverter
    Implements IValueConverter
    Private rankList As Ranks
    Public Sub New()
      Ranks.GetRanks(AddressOf EndGetRanks)
    End Sub

    Private Sub EndGetRanks(ByVal sender As Object, ByVal e As Csla.DataPortalResult(Of Ranks))
      rankList = e.Object
      OnGotData()
    End Sub


    Public Event GotData As EventHandler

    Protected Sub OnGotData()
      If GotDataEvent IsNot Nothing Then
        GotDataEvent.Invoke(Me, EventArgs.Empty)
      End If
    End Sub

    Public Function Convert(ByVal value As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As System.Globalization.CultureInfo) As Object Implements IValueConverter.Convert
      If value IsNot Nothing AndAlso TypeOf value Is Integer AndAlso rankList IsNot Nothing Then
        Return rankList.GetRankName(CInt(Fix(value)))
      Else
        Return String.Empty
      End If

    End Function

    Public Function ConvertBack(ByVal value As Object, ByVal targetType As Type, ByVal parameter As Object, ByVal culture As System.Globalization.CultureInfo) As Object Implements IValueConverter.ConvertBack
      If value IsNot Nothing AndAlso TypeOf value Is String AndAlso rankList IsNot Nothing Then
        Dim returnValue = rankList.GetItemByValue(CStr(value))
        If returnValue IsNot Nothing Then
          Return returnValue.Key
        Else
          Return 0
        End If
      Else
        Return 0
      End If
    End Function
  End Class
End Namespace
